DSpace/dspace-angular
# Features support part 2

## References
This PR expands on #717
and fixes #738

## Description
This PR hides / protects certain components and routes behind feature authorization checks. Support for this was added in #717, this PR aims to cover more features using these changes.

## Instructions for Reviewers
Changes made:
- Implemented abstract `DsoPageFeatureGuard`, accepting a resolver (resolving the route's ID to a remote-data object). `getObjectUrl` returns the self-link of the object resolved from the URL, which will be used for checking the Feature's authorization
- Implementations of this abstract guard:
-- `CollectionPageAdministratorGuard`: checks admin authorization for collection pages. Configured for `/collection/${uuid}/edit` routes.
-- `CommunityPageAdministratorGuard`: same, but for communities
-- `ItemPageAdministratorGuard`: same, but for items
-- `ItemPageReinstateGuard`: checks "reinstateitem" feature authorization for item pages. Configured for `/items/${uuid}/edit/reinstate` routes.
-- `ItemPageWithdrawGuard`: checks "withdrawitem" feature authorization for item pages. Configured for `/items/${uuid}/edit/withdraw` routes.
- Created `SiteRegisterGuard` checking "epersonRegistration" feature authorization for the Site. Configured for `/register` route.
- Added checks for the register link (in the login dropdown) to be hidden when not authorized.
- Added checks for the withdraw or reinstate button (on the edit item page) to be hidden when not authorized.

How to test:
- Visiting the mentioned pages containing links/buttons while logged in as an unauthorized user, should not display these links/buttons.
- Trying to visit the mentioned routes with an unauthorized should result in the user ending up on the `/unauthorized` page.

**Warnings**
- [#2912](https://github.com/DSpace/DSpace/issues/2912): This issue might cause the administratorOf feature to not work as expected. Take this into account when testing this PR.
- #635: Authorization requests will still be cached for the previous user when logging in (usually anonymous), which causes some stuff like the sidebar to not update accordingly. Also take this into account when testing this PR.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
